[UnifiedPDF] Context menu item "Previous Page" appears on the first page in Two Page view layouts

aprotyas opened the pull request:
#### b98c84d7c4a00d2a516edec6ee696695724087f2
<pre>
[UnifiedPDF] Context menu item &quot;Previous Page&quot; appears on the first page in Two Page view layouts
<a href="https://bugs.webkit.org/show_bug.cgi?id=273745">https://bugs.webkit.org/show_bug.cgi?id=273745</a>
<a href="https://rdar.apple.com/127442932">rdar://127442932</a>

Reviewed by Simon Fraser.

This patch makes the logic to mark the &quot;Previous/Next Page&quot; context menu
items as active more versatile. It does so by:

1. Passing the page index where the context menu event was fired to
   navigationContextMenuItems, instead of trying to infer the view&apos;s page
   index through indexForCurrentPageInView.

2. Taking into account the fact that for 2up display modes, we want the
   first pair of pages to have an inactive &quot;Previous Page&quot; item.

3. Computing an effective last page index, below which the &quot;Next Page&quot;
   context menu item is active for all pages.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::createContextMenu const):
(WebKit::UnifiedPDFPlugin::navigationContextMenuItemsForPageAtIndex const):
(WebKit::UnifiedPDFPlugin::navigationContextMenuItems const): Deleted.

Canonical link: <a href="https://commits.webkit.org/278417@main">https://commits.webkit.org/278417@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c328c87a831847d99862ab198174abfc55ba599b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50493 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29789 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2797 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/53752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1183 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36037 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/833 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/53752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52592 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/27443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43468 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/53752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/24851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8871 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/46833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/787 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55341 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25591 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/48590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/26852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/43623 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47636 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11064 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27716 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/26584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2877,12 +2881,18 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
 }
 
-Vector<PDFContextMenuItem> UnifiedPDFPlugin::navigationContextMenuItems() const
+Vector<PDFContextMenuItem> UnifiedPDFPlugin::navigationContextMenuItems(PDFDocumentLayout::PageIndex contextMenuEventPage) const
```
You could rename this `navigationContextMenuItemsForPageAtIndex()` and then the arg could just be `pageIndex`

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2877,12 +2881,18 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+Vector<PDFContextMenuItem> UnifiedPDFPlugin::navigationContextMenuItems(PDFDocumentLayout::PageIndex contextMenuEventPage) const
 {
-    auto currentPageIndex = indexForCurrentPageInView();
+    PDFDocumentLayout::PageIndex pageIncrement = isShowingTwoPages() ? 2 : 1;
```
Could use `pagesPerRow` in `PDFDocumentLayout`

aprotyas committed:
[UnifiedPDF] Context menu item "Previous Page" appears on the first page in Two Page view layouts
https://bugs.webkit.org/show_bug.cgi?id=273745
rdar://127442932

Reviewed by Simon Fraser.

This patch makes the logic to mark the "Previous/Next Page" context menu
items as active more versatile. It does so by:

1. Passing the page index where the context menu event was fired to
   navigationContextMenuItems, instead of trying to infer the view's page
   index through indexForCurrentPageInView.

2. Taking into account the fact that for 2up display modes, we want the
   first pair of pages to have an inactive "Previous Page" item.

3. Computing an effective last page index, below which the "Next Page"
   context menu item is active for all pages.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::createContextMenu const):
(WebKit::UnifiedPDFPlugin::navigationContextMenuItemsForPageAtIndex const):
(WebKit::UnifiedPDFPlugin::navigationContextMenuItems const): Deleted.

Canonical link: https://commits.webkit.org/278417@main
